system_test/aest_sync_SUITE.erl

```diff
@@ -0,0 +1,190 @@
+    timer:sleep((Length div BlocksPerSecond) * 1000),
+
+    wait_for_height(Length, [node3], 1000, Cfg),
+    ct:log("Node 3 on same height"),
+    Height = request(node3, [v2, 'block-by-height'], #{height => Length}, Cfg),
+    ok.
+
+%% When we stop and restart a node we will be able to read the blocks
+%% that we had in the chain before stopping: data is persistent.
+docker_keeps_data(Cfg) ->
+    Length = 20,
+    NodeStartupTime = proplists:get_value(node_startup_time, Cfg),
+
+    setup_nodes([?NODE1], Cfg),
+    start_node(node1, Cfg),
+    wait_for_height(0, [node1], NodeStartupTime, Cfg),
+
+    %% Mine for 20 blocks and calculate the average mining time
+    StartTime = os:timestamp(),
+    wait_for_height(Length, [node1], Length * ?MINING_TIMEOUT, Cfg),
+    EndTime = os:timestamp(),
+    %% Average mining time per block plus 50% extra
+    MiningTime = round(timer:now_diff(EndTime, StartTime) * 1.5)
+                 div (1000 * Length),
+
+    %% Get all blocks before stopping
+    A = [get_block(node1, H, Cfg) || H <- lists:seq(1, Length)],
+
+    stop_node(node1, infinity, Cfg), %% Is this triggering PT-155851463 ?
+    start_node(node1, Cfg),
+    wait_for_height(0, [node1], NodeStartupTime, Cfg),
+
+    ct:log("Node restarted and ready to go"),
+
+    %% Give it time to read from disk, but not enough to build a new chain of same length
+    timer:sleep(MiningTime * 3),
+
+    %% Get all blocks after restarting
+    B = [get_block(node1, H, Cfg) || H <- lists:seq(1, Length)],
+
+    {_, Diff} = lists:foldl(fun({X, Y}, {H, Acc}) ->
+        case X =:= Y of
+            true -> {H + 1, Acc};
+            false ->
+                ct:log("Block ~w changed after restart:~n"
+                       "BEFORE:~n~p~nAFTER:~n~p~n", [H, X, Y]),
+                {H + 1, [H | Acc]}
+        end
+    end, {1, []}, lists:zip(A, B)),
+    ?assertEqual([], Diff),
+
+    ok.
+
+crash_and_continue_sync(Cfg) ->
+    BlocksPerSecond = proplists:get_value(blocks_per_second, Cfg),
+    NodeStartupTime = proplists:get_value(node_startup_time, Cfg),
+    NodeStopTime = proplists:get_value(node_stop_time, Cfg),
+    %% Create a chain long enough to need 10 seconds to fetch it
+    Length = BlocksPerSecond * 10,
+
+    setup_nodes([?NODE1, ?NODE2], Cfg),
+    start_node(node1, Cfg),
+    wait_for_height(Length, [node1], Length * ?MINING_TIMEOUT, Cfg),
+    B1 = request(node1, [v2, 'block-by-height'], #{height => Length}, Cfg),
+    ct:log("Node 1 at height ~p: ~p~n", [Length, B1]),
+
+    %% Start fetching the chain
+    start_node(node2, Cfg),
+    wait_for_height(0, [node2], NodeStartupTime, Cfg),
+    ct:log("Node 2 ready to go"),
+
+    %% we are fetching blocks crash now
+    stop_node(node1, NodeStopTime div 1000, Cfg),
+    Top2 = request(node2, [v2, 'top'], #{}, Cfg),
+    ct:log("Node 2 top: ~p~n", [Top2]),
+    Height = maps:get(height, Top2),
+    case Height >= Length of
+         true -> {skip, already_synced_when_crashed};
```

Comment:
Shouldn't this be more deterministic? In some cases we might not be testing anything at all?

Reply:
This question would be for @ThomasArts , I didn't look much at this test.